qutip/tests/core/data/test_reshape.py

```diff
@@ -0,0 +1,135 @@
+        (pytest.param((10, 10), id="square"), ),
+        (pytest.param((2, 10), id="non_square"), ),
+    ]
+
+    specialisations = [
+        pytest.param(data.column_unstack_csr, CSR, CSR),
+        pytest.param(data.column_unstack_dense, Dense, Dense),
+    ]
+
+    # `ColumnUnstack` has an additional scalar parameters. This is why tests
+    # are a little bit more involved.
+    @pytest.mark.parametrize('rows', [2, 5])
+    def test_mathematically_correct(self, op, data_m, rows, out_type):
+        """
+        Test that the binary operation is correct for all the known type
+        specialisations.
+        """
+        matrix = data_m()
+        expected = self.op_numpy(matrix.to_array(), rows)
+        test = op(matrix, rows)
+        assert isinstance(test, out_type)
+        assert test.shape == expected.shape
+        np.testing.assert_allclose(test.to_array(), expected, atol=self.tol)
+
+    # These tests will check that ValueError is raised when input matrices have
+    # invalid shape.
+    def test_incorrect_shape_raises(self, op, data_m):
+        with pytest.raises(ValueError):
+            op(data_m(), 1)  # We set rows to one so that it is always valid
+
+    # We also test that wrong value for `rows` raises ValueError We will
+    # generate tests with valid data_m since we want to make sure that the
+    # error is raised due to wrong `rows` value. For this we will make use of
+    # the generate_mathematically_correct function. This is why out_type is
+    # included although not used at all.
+    @pytest.mark.parametrize('rows', [-1, 0, 3], ids=['negative', 'zero',
+                                                      'invalid'])
+    def test_incorrect_rows_raises(self, op, data_m, out_type, rows):
+        with pytest.raises(ValueError):
+            op(data_m(), rows)
+
+    # To generate tests we simply call to getenerate_mathematically correct.
+    def generate_incorrect_rows_raises(self, metafunc):
+        self.generate_mathematically_correct(metafunc)
+
+
+class TestReshape(UnaryOpMixin):
+    def op_numpy(self, matrix, rows, columns):
+        # It seem to_array returns F ordered numpy arrays.
+        out_shape = (rows, columns)
+        return np.reshape(matrix, newshape=out_shape, order='C')
+
+    # I use custom shapes so that they all have 100 elements.
+    shapes = [
+        (pytest.param((1, 100), id="bra"), ),
+        (pytest.param((100, 1), id="ket"), ),
+        (pytest.param((2, 50), id="non_square"), ),
+        (pytest.param((10, 10), id="square"), ),
+    ]
+
+    specialisations = [
+        pytest.param(data.reshape_dense, Dense, Dense),
+        pytest.param(data.reshape_csr, CSR, CSR),
+    ]
+
+    # `reshape` has an additional scalar parameters. This is why tests
+    # are a little bit more involved.
+    @pytest.mark.parametrize('rows, columns', [(5, 20), (10, 10)])
+    def test_mathematically_correct(self, op, data_m, rows, columns, out_type):
```

Comment:
Why parametrise over two parameters when you could just parametrise over `new_shape`?

Reply:
So that the parameters match the required parameters for the reshape specialisation.

Reply:
Yeah, fair enough. It looks a little odd to me to parametrise over them rather than a single linked variable "shape", but that's ok.